File: examples/contrib/divisible_by_9_through_1.py
# Copyright 2010 Hakan Kjellerstrand hakank@gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Divisible by 9 through 1 puzzle in Google CP Solver.

From http://msdn.microsoft.com/en-us/vcsharp/ee957404.aspx
' Solving Combinatory Problems with LINQ'
'''
Find a number consisting of 9 digits in which each of the digits
from 1 to 9 appears only once. This number must also satisfy these
divisibility requirements:

 1. The number should be divisible by 9.
 2. If the rightmost digit is removed, the remaining number should
    be divisible by 8.
 3. If the rightmost digit of the new number is removed, the remaining
    number should be divisible by 7.
 4. And so on, until there's only one digit (which will necessarily
    be divisible by 1).
'''

Also, see
'Intel Parallel Studio: Great for Serial Code Too (Episode 1)'
http://software.intel.com/en-us/blogs/2009/12/07/intel-parallel-studio-great-for-serial-code-too-episode-1/


This model is however generalized to handle any base, for reasonable limits.
The 'reasonable limit' for this model is that base must be between 2..16.

Compare with the following models:
* MiniZinc: http://www.hakank.org/minizinc/divisible_by_9_through_1.mzn
* Comet   : http://www.hakank.org/comet/divisible_by_9_through_1.co
* ECLiPSe : http://www.hakank.org/eclipse/divisible_by_9_through_1.ecl
* Gecode  : http://www.hakank.org/gecode/divisible_by_9_through_1.cpp


This model was created by Hakan Kjellerstrand (hakank@gmail.com)
Also see my other Google CP Solver models:
http://www.hakank.org/google_or_tools/
"""

import sys

from ortools.constraint_solver.python import constraint_solver as cp

#
# Decomposition of modulo constraint
#
# This implementation is based on the ECLiPSe version
# mentioned in
# - A Modulo propagator for ECLiPSE'
#   http://www.hakank.org/constraint_programming_blog/2010/05/a_modulo_propagator_for_eclips.html
# The ECLiPSe source code:
# http://www.hakank.org/eclipse/modulo_propagator.ecl
#


def my_mod(solver, x, y, r):

    if not isinstance(y, int):
        solver.add(y != 0)

    lbx = x.min()
    ubx = x.max()
    ubx_neg = -ubx
    lbx_neg = -lbx
    min_x = min(lbx, ubx_neg)
    max_x = max(ubx, lbx_neg)

    d = solver.new_int_var(max(0, min_x), max_x, "d")

    if not isinstance(r, int):
        solver.add(r >= 0)
        solver.add(x * r >= 0)

    if not isinstance(r, int) and not isinstance(r, int):
        solver.add(-abs(y) < r)
        solver.add(r < abs(y))

    solver.add(min_x <= d)
    solver.add(d <= max_x)
    solver.add(x == y * d + r)


#
# converts a number (s) <-> an array of integers (t) in the specific base.
#
def toNum(solver, t, s, base):
    tlen = len(t)
    solver.add(s == solver.sum([(base ** (tlen - i - 1)) * t[i] for i in range(tlen)]))


def main(base=10):

    # Create the solver.
    solver = cp.Solver("Divisible by 9 through 1")

    # data
    m = base ** (base - 1) - 1
    n = base - 1

    digits_str = "_0123456789ABCDEFGH"

    print("base:", base)

    # declare variables

    # the digits
    x = [solver.new_int_var(1, base - 1, "x[%i]" % i) for i in range(n)]

    # the numbers, t[0] contains the answer
    t = [solver.new_int_var(0, m, "t[%i]" % i) for i in range(n)]

    #
    # constraints
    #
    solver.add_all_different(x)

    for i in range(n):
        mm = base - i - 1
        toNum(solver, [x[j] for j in range(mm)], t[i], base)
        my_mod(solver, t[i], mm, 0)

    #
    # solution and search
    #
    solution = solver.assignment()
    solution.add(x)
    solution.add(t)

    db = solver.phase(
        x,
        cp.IntVarStrategy.CHOOSE_FIRST_UNBOUND,
        cp.IntValueStrategy.ASSIGN_MIN_VALUE,
    )

    solver.new_search(db)
    num_solutions = 0
    while solver.next_solution():
        print("x: ", [x[i].value() for i in range(n)])
        print("t: ", [t[i].value() for i in range(n)])
        print(
            "number base 10: %i base %i: %s"
            % (
                t[0].value(),
                base,
                "".join([digits_str[x[i].value() + 1] for i in range(n)]),
            )
        )
        print()
        num_solutions += 1
    solver.end_search()

    print("num_solutions:", num_solutions)
    print("failures:", solver.num_failures)
    print("branches:", solver.num_branches)
    print("WallTime:", solver.wall_time_ms)


base = 10
default_base = 10
max_base = 16
if __name__ == "__main__":
    if len(sys.argv) > 1:
        base = int(sys.argv[1])
        if base > max_base:
            print(
                "Sorry, max allowed base is %i. Setting base to %i..."
                % (max_base, default_base)
            )
            base = default_base
    main(base)

    # for base in range(2, 17):
    #     print
    #     main(base)

File: examples/contrib/divisible_by_9_through_1.cs
//
// Copyright 2012 Hakan Kjellerstrand
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections;
using System.IO;
using System.Text.RegularExpressions;
using Google.OrTools.ConstraintSolver;

public class DivisibleBy9Through1
{
    /**
     *
     * A simple propagator for modulo constraint.
     *
     * This implementation is based on the ECLiPSe version
     * mentioned in "A Modulo propagator for ECLiPSE"
     * http://www.hakank.org/constraint_programming_blog/2010/05/a_modulo_propagator_for_eclips.html
     * The ECLiPSe Prolog source code:
     * http://www.hakank.org/eclipse/modulo_propagator.ecl
     *
     */
    public static void MyMod(Solver solver, IntVar x, IntVar y, IntVar r)
    {
        long lbx = x.Min();
        long ubx = x.Max();
        long ubx_neg = -ubx;
        long lbx_neg = -lbx;
        int min_x = (int)Math.Min(lbx, ubx_neg);
        int max_x = (int)Math.Max(ubx, lbx_neg);

        IntVar d = solver.MakeIntVar(min_x, max_x, "d");

        // r >= 0
        solver.Add(r >= 0);

        // x*r >= 0
        solver.Add(x * r >= 0);

        // -abs(y) < r
        solver.Add(-y.Abs() < r);

        // r < abs(y)
        solver.Add(r < y.Abs());

        // min_x <= d, i.e. d > min_x
        solver.Add(d > min_x);

        // d <= max_x
        solver.Add(d <= max_x);

        // x == y*d+r
        solver.Add(x - (y * d + r) == 0);
    }

    /**
     *
     *  ToNum(solver, a, num, base)
     *
     *  channelling between the array a and the number num
     *
     */
    private static Constraint ToNum(IntVar[] a, IntVar num, int bbase)
    {
        int len = a.Length;

        IntVar[] tmp = new IntVar[len];
        for (int i = 0; i < len; i++)
        {
            tmp[i] = (a[i] * (int)Math.Pow(bbase, (len - i - 1))).Var();
        }
        return tmp.Sum() == num;
    }

    /**
     *
     * Solves the divisible by 9 through 1 problem.
     * See http://www.hakank.org/google_or_tools/divisible_by_9_through_1.py
     *
     */
    private static void Solve(int bbase)
    {
        Solver solver = new Solver("DivisibleBy9Through1");

        int m = (int)Math.Pow(bbase, (bbase - 1)) - 1;
        int n = bbase - 1;

        String[] digits_str = { "_", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9" };

        Console.WriteLine("base: " + bbase);

        //
        // Decision variables
        //
        // digits
        IntVar[] x = solver.MakeIntVarArray(n, 1, bbase - 1, "x");

        // the numbers. t[0] contains the answe
        IntVar[] t = solver.MakeIntVarArray(n, 0, m, "t");

        //
        // Constraints
        //

        solver.Add(x.AllDifferent());

        // Ensure the divisibility of base .. 1
        IntVar zero = solver.MakeIntConst(0);
        for (int i = 0; i < n; i++)
        {
            int mm = bbase - i - 1;
            IntVar[] tt = new IntVar[mm];
            for (int j = 0; j < mm; j++)
            {
                tt[j] = x[j];
            }
            solver.Add(ToNum(tt, t[i], bbase));
            MyMod(solver, t[i], solver.MakeIntConst(mm), zero);
        }

        //
        // Search
        //
        DecisionBuilder db = solver.MakePhase(x, Solver.INT_VAR_DEFAULT, Solver.INT_VALUE_DEFAULT);

        solver.NewSearch(db);

        while (solver.NextSolution())
        {
            Console.Write("x: ");
            for (int i = 0; i < n; i++)
            {
                Console.Write(x[i].Value() + " ");
            }
            Console.WriteLine("\nt: ");
            for (int i = 0; i < n; i++)
            {
                Console.Write(t[i].Value() + " ");
            }
            Console.WriteLine("\n");

            if (bbase != 10)
            {
                Console.Write("Number base 10: " + t[0].Value());
                Console.Write(" Base " + bbase + ": ");
                for (int i = 0; i < n; i++)
                {
                    Console.Write(digits_str[(int)x[i].Value() + 1]);
                }
                Console.WriteLine("\n");
            }
        }

        Console.WriteLine("\nSolutions: {0}", solver.Solutions());
        Console.WriteLine("WallTime: {0}ms", solver.WallTime());
        Console.WriteLine("Failures: {0}", solver.Failures());
        Console.WriteLine("Branches: {0} ", solver.Branches());

        solver.EndSearch();
    }

    public static void Main(String[] args)
    {
        int bbase = 10;
        if (args.Length > 0)
        {
            bbase = Convert.ToInt32(args[0]);
            if (bbase > 12)
            {
                // Though base = 12 has no solution...
                Console.WriteLine("Sorry, max relevant base is 12. Setting base to 12.");
                bbase = 10;
            }
        }

        Solve(bbase);
    }
}
